libp2p/utils/semaphore.nim

```diff
@@ -77,13 +78,14 @@ proc release*(s: AsyncSemaphore) =
     trace "Releasing slot", available = s.count,
                             queue = s.queue.len
 
+    s.count.inc
     if s.queue.len > 0:
       var fut = s.queue[0]
       s.queue.delete(0)
       if not fut.finished():
```

Comment:
this code looks odd btw: why would there be finished futures in the queue?

Reply:
and as a follow-up, if a future is finished already, should the "next" future be completed instead?

Reply:
A future could be cancelled
Not sure why we don't try another, I switched to a loop and every tests pass, @dryajov?

Reply:
the cancelled futures remove themselves from the queue in their cancellation callback, so that _should_ never happen..
what could happen is that the user "manually" completes the future but that's not how one uses futures in general so..

Reply:
Right, missed that part

Reply:
if we can't figure out a reason, might be a good idea to assert with a message that the future was unexpectedly completed - by extension, a loop would no longer be needed - if this happens, I get the feeling we might get an imbalance in acquires and releases which either leads to lockups or overallocation

Reply:
Since the counter and the queue size can now diverge due to the `forceAcquire` contraption, we might as well have to use a loop in `release`.